rfcs/0051-hard-fork-package-generation.md

```diff
@@ -0,0 +1,148 @@
+
+### Generating the Hard Fork Package
+
+The hard fork package will be generated in a buildkite pipeline. The buildkite pipeline will accept environment variables providing URLs to download the unmigrated genesis state from. Because different jobs require different parts of the genesis state to run, we will split up the unmigrated genesis state inputs into the following files:
+
+- `block.json` (the full protocol state of the dumped block)
+- `staged_ledger.json` (the contents of the staged ledger of the dumped block)
+- `next_staking_ledger.json` (the contents of the next staking ledger of the dumped block)
+- `staking_ledger.json` (the contents of the staking ledger of the dumped block)
+
+Here is a diagram of the various buildkite jobs required to generate the hard fork package:
+
+![](res/hard-fork-package-generation-buildkite-pipeline.dot.png)
+
+The `build_daemon` job will build the daemon with the correct compile time configuration for mainnet (`mainnet.mlh`). The hard fork specific configuration will be provided via a static runtime configuration that is bundled with the hard fork release.
+
+The `generate_network_config` job will generate the runtime configuration for the hard fork network, embedding the correct ledger hashes and correct fields for `fork_previous_length`, `fork_previous_state_hash` and `fork_previous_global_slot`.
+
+* IMPORTANT: when does the genesis\_timestamp get written in? has is it configured?
+
+The `migrate_*_ledger` jobs will perform any necessary migrations on the input ledgers, and will generate rocksdb ledgers that will be packaged with the hard fork release. These jobs run in parallel.
+
+The `package_deb` and `package_dockerhub` jobs are our standard artifact bundling jobs that we use for releases, with some modifications to pull in the correct data from the hard fork build pipeline. The `package_deb` step will bundle together that artifacts from all the prior steps into a `.deb` package. The `package_dockerhub` installs this `.deb` package onto a Ubuntu docker image.
+
+In the future, we will add additional jobs to the pipeline that will also perform automated tests against the release candidate.
+
+### Testing the Hard Fork Package
+
+In the future, we wish to automate the process of testing a hard fork release candidate. However for the upcoming hard fork, we will instead rely on a manual testing process in order to verify that the hard fork release candidate is of sufficient quality to release. The goal of this testing at this layer is to determine that the new package is configured to the proper genesis state, and that a network can be initialized using the package. We do not perform wider scope functional test suites at this stage, as the version of the software being built when deploying a hard fork will have already been tested thoroughly; we are only interested in testing workflows introduced or affected by the hard fork package generation process.
```

Comment:
That would be a good test, but isn't the agreed upon scope for ITN. I think changing the plan for that network this late isn't a good idea. However, we should (and had planned to) run a test like this on a separate network. In addition, we do plan to test a hard fork on the ITN network, but it will be between versions of Berkeley rather than from mainnet to Berkeley.